prog/dftb+/prg_dftb/mainio.F90

```diff
@@ -82,524 +96,1459 @@ module mainio
+    if (tPrintEigvecsTxt) then
+      if (denseDesc%t2Component) then
+        call writePauliEigvecsTxtSerial(fd, neighborList, nNeighbor, denseDesc, iPair,&
+            & img2CentCell, iCellVec, cellVec, orb, species, speciesName, over, parallelKS, kPoint,&
+            & eigvecsCplx, SSqrCplx, fileName=fileName)
+      else
+        call writeCplxEigvecsTxtSerial(fd, neighborList, nNeighbor, denseDesc, iPair, img2CentCell,&
+            & iCellVec, cellVec, orb, species, speciesName, over, parallelKS, kPoint, eigvecsCplx,&
+            & SSqrCplx, fileName=fileName)
+      end if
+    end if
+  #:endif
+
+  end subroutine writeCplxEigvecs
+
+
+#:for DTYPE, NAME in [('complex', 'Cplx'), ('real', 'Real')]
+
+#:if WITH_SCALAPACK
+
+  !> Write the real eigvectors into binary output file (BLACS version).
+  subroutine write${NAME}$EigvecsBinBlacs(env, denseDesc, eigvecs, fd, runId, parallelKS, fileName)
+
+    !> Environment settings
+    type(TEnvironment), intent(in) :: env
+
+    !> Dense matrix descriptor
+    type(TDenseDescr), intent(in) :: denseDesc
 
     !> Square Hamiltonian (or work array)
-    complex(dp), intent(inout) :: HSqrCplx(:,:,:,:)
+    ${DTYPE}$(dp), intent(in) :: eigvecs(:,:,:)
 
-    !> Work array.
-    complex(dp), intent(inout) :: SSqrCplx(:,:)
+    !> Fileid (file not yet opened) to use.
+    integer, intent(in) :: fd
+
+    !> Id of the current program run.
+    integer, intent(in) :: runId
 
-    !> If present, Hamiltonian(s) are fetched from this storage into HSqrCplx, instead of using
-    !> whatever is already there.
-    type(OFifoCplxR2), intent(inout), optional :: storeEigvecs(:)
+    !> K-points and spins to process
+    type(TParallelKS), intent(in) :: parallelKS
 
     !> optional alternative file prefix, to appear as "fileName".bin
     character(len=*), intent(in), optional :: fileName
 
-    character(lc) :: tmpStr
-    integer :: iSpin, iSpin2, iAtom, iSp1, iSh1, iOrb, ang, iK, iK2, nK
-    integer :: ii, jj, nOrb, nSpinChannel
-    complex(dp), allocatable :: cVecTemp(:), work(:,:)
+    type(linecomm) :: collector
+    ${DTYPE}$(dp), allocatable :: localEigvec(:)
+    integer :: nOrb
+    integer :: iKS, iGroup, iEig
 
-    nK = size(kPoint, dim=2)
-    close(fdEigvec) ! just to be on the safe side
-    if (present(fileName)) then
-      write (tmpStr, "(A,A)") trim(fileName), ".bin"
-      open(fdEigvec, file=tmpStr, action="write", status="replace", &
-          &position="rewind", form="unformatted")
-    else
-      open(fdEigvec, file=eigvecBin, action="write", status="replace", &
-          &position="rewind", form="unformatted")
+    nOrb = denseDesc%fullSize
+    allocate(localEigvec(nOrb))
+
+    if (env%mpi%tGlobalMaster) then
+      call prepareEigvecFileBin(fd, runId, fileName)
     end if
-    write (fdEigVec) runId
-    if (nSpin == 4) then
-      nSpinChannel = 1
+    call collector%init(env%blacs%orbitalGrid, denseDesc%blacsOrbSqr, "c")
+
```

Comment:
Might be useful to comment the communication between groups and the interaction with the iKS loops, as well as the collector calls, as some of the communication logic in later cases isn't obvious on first inspection.